src/IO/S3/Client.h

```diff
@@ -0,0 +1,309 @@
+        {
+            ClientCacheRegistry::instance().unregisterClient(cache.get());
+        }
+        catch (...)
+        {
+            tryLogCurrentException(log);
+            throw;
+        }
+    }
+
+    /// Decorator for RetryStrategy needed for this client to work correctly
+    class RetryStrategy : public Aws::Client::RetryStrategy
+    {
+    public:
+        explicit RetryStrategy(std::shared_ptr<Aws::Client::RetryStrategy> wrapped_strategy_);
+
+        /// NOLINTNEXTLINE(google-runtime-int)
+        bool ShouldRetry(const Aws::Client::AWSError<Aws::Client::CoreErrors>& error, long attemptedRetries) const override;
+
+        /// NOLINTNEXTLINE(google-runtime-int)
+        long CalculateDelayBeforeNextRetry(const Aws::Client::AWSError<Aws::Client::CoreErrors>& error, long attemptedRetries) const override;
+
+        /// NOLINTNEXTLINE(google-runtime-int)
+        long GetMaxAttempts() const override;
+
+        void GetSendToken() override;
+
+        bool HasSendToken() override;
+
+        void RequestBookkeeping(const Aws::Client::HttpResponseOutcome& httpResponseOutcome) override;
+        void RequestBookkeeping(const Aws::Client::HttpResponseOutcome& httpResponseOutcome, const Aws::Client::AWSError<Aws::Client::CoreErrors>& lastError) override;
+    private:
+        std::shared_ptr<Aws::Client::RetryStrategy> wrapped_strategy;
+    };
+
+    Model::HeadObjectOutcome HeadObject(const HeadObjectRequest & request) const;
+    Model::ListObjectsV2Outcome ListObjectsV2(const ListObjectsV2Request & request) const;
+    Model::ListObjectsOutcome ListObjects(const ListObjectsRequest & request) const;
+    Model::GetObjectOutcome GetObject(const GetObjectRequest & request) const;
+
+    Model::AbortMultipartUploadOutcome AbortMultipartUpload(const AbortMultipartUploadRequest & request) const;
+    Model::CreateMultipartUploadOutcome CreateMultipartUpload(const CreateMultipartUploadRequest & request) const;
+    Model::CompleteMultipartUploadOutcome CompleteMultipartUpload(const CompleteMultipartUploadRequest & request) const;
+    Model::UploadPartOutcome UploadPart(const UploadPartRequest & request) const;
+    Model::UploadPartCopyOutcome UploadPartCopy(const UploadPartCopyRequest & request) const;
+
+    Model::CopyObjectOutcome CopyObject(const CopyObjectRequest & request) const;
+    Model::PutObjectOutcome PutObject(const PutObjectRequest & request) const;
+    Model::DeleteObjectOutcome DeleteObject(const DeleteObjectRequest & request) const;
+    Model::DeleteObjectsOutcome DeleteObjects(const DeleteObjectsRequest & request) const;
+
+private:
+    template <typename... Args>
+    explicit Client(size_t max_redirects_, Args &&... args)
+        : Aws::S3::S3Client(std::forward<Args>(args)...)
+        , max_redirects(max_redirects_)
+        , log(&Poco::Logger::get("S3Client"))
+    {
+        auto * endpoint_provider = dynamic_cast<Aws::S3::Endpoint::S3DefaultEpProviderBase *>(accessEndpointProvider().get());
```

Comment:
Can `endpoint_provider` be nullptr?

Reply:
it shouldn't be, AWS always does a nullptr check before accessing it, I can do `chassert`.